### PR TITLE
8315702: jcmd Thread.dump_to_file slow with millions of virtual threads

### PR DESCRIPTION
**Pull request by AlanBateman**
`HotSpotDiagnosticMXBean.dumpThreads` and `jcmd Thread.dump_to_file` are slow when there is a large number of threads.

The thread dump can be sped up significantly with some small changes:
- Using println rather than format when printing the thread info and thread stacks
- Wrap the underlying file output stream in a BufferedOutputStream

With 200k virtual threads on macOS, the plain format thread dump goes from 22s to 1.8s, and the json format thread dump goes from 31s to 2.8s on one system that I tried. On a Linux system, also with 200k threads, the plain thread dump goes from 8.7s to 2.9s, and the json format thread dump from 12.4s to 4.5s.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315702](https://bugs.openjdk.org/browse/JDK-8315702): jcmd Thread.dump_to_file slow with millions of virtual threads (**Bug** - P4)

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15581/head:pull/15581` \
`$ git checkout pull/15581`

Update a local copy of the PR: \
`$ git checkout pull/15581` \
`$ git pull https://git.openjdk.org/jdk.git pull/15581/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15581`

View PR using the GUI difftool: \
`$ git pr show -t 15581`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15581.diff">https://git.openjdk.org/jdk/pull/15581.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15581#issuecomment-1709999382)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/vm/ThreadDumper.java [hunk 1]
```diff
@@ -158,9 +160,9 @@ private static void dumpThreads(ThreadContainer container, PrintStream ps) {
 
     private static void dumpThread(Thread thread, PrintStream ps) {
         String suffix = thread.isVirtual() ? " virtual" : "";
-        ps.format("#%d \"%s\"%s%n", thread.threadId(), thread.getName(), suffix);
+        ps.println("#" + thread.threadId() + " \"" + thread.getName() + "\"" + suffix);
         for (StackTraceElement ste : thread.getStackTrace()) {
-            ps.format("      %s%n", ste);
+            ps.println("      " + ste);
```
Comment (schlosna) [inline src/java.base/share/classes/jdk/internal/vm/ThreadDumper.java:165]: I'm curious if avoiding the string concat is more or less efficient for the volume of virtual threads and stack trace elements being written, though it does open the window to interleaved writes to the underlying stream.

```suggestion
            ps.print("      ");
            ps.println(ste);
```

**Comment by schlosna** [comment]
I don't think more counts, but looks good to me.

**Review by mlchung — APPROVED** [review]
Looks good to me.

**Review by alexmenkov — APPROVED** [review]

**Review by jaikiran — APPROVED** [review]
This looks good to me.

**Comment by AlanBateman** [comment]
Thanks for reviews.

**Pull request closed without merging**